### DISCUSSION TITLE
Kepler.GL integration in Next JS app

### DISCUSSION THREAD

**Discussion by SusantaKumar27** [vercel/next.js#36468] [Help] [answered]
I am integrating Kepler.GL in my Next JS app, but while rendering the <KeplerGL /> component, I am getting the following error.

`Error: Could not find "store" in the context of "Connect(Container)". Either wrap the root component in a <Provider> or pass a custom React context provider to <Provider> and the corresponding React context consumer to Connect(Container) in connect options.`

I have tried different versions of the packages, as well as in different Next JS apps. Also, I have wrapped my root component in a <Provider /> tag and passed the store correctly (as suggested in the error), but still getting the same error every time.

 - store.js
```import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';
import { createStore, combineReducers, applyMiddleware } from 'redux';
import keplerGlReducer from 'kepler.gl/reducers';
import { appReducer } from '../reducers/appReducer';

const initialState = {};

const reducers = combineReducers({
    app: appReducer,
    keplerGl: keplerGlReducer,
});
const middleware = [thunk];

export const store = createStore(
    reducers,
    initialState,
    composeWithDevTools(applyMiddleware(...middleware))
);
```

 - app.js
```import '../styles/globals.css';
import { Provider } from 'react-redux';
import { store } from '../store';

function MyApp({ Component, pageProps }) {
    return (
        <Provider store={store}>
            <Component {...pageProps} />
        </Provider>
    );
}

export default MyApp;
```

 - KeplerMap.js
```import React from 'react';
import KeplerGl from 'kepler.gl';

function KeplerMap(props) {
    return (
        <div>
            <KeplerGl
                id="example"
                mapboxApiAccessToken={process.env.REACT_APP_MAPBOX_TOKEN}
                width={400}
                height={400}
            />
        </div>
    );
}

export default KeplerMap;
```

I am not sure that Kepler.GL is compatible with Next JS or not, but I want to know why this error is showing and how to resolve it

**Comment by icyJoseph** [comment, marked as answer]
Hi, 

I think you are missing the `Kepler.GL` middleware, `import {enhanceReduxMiddleware} from 'kepler.gl/middleware';`, but funnily enough, when I include it I get other errors. I found it in the [docs](https://github.com/keplergl/kepler.gl#1-mount-reducer) by the way.

In your case do:

```js
const middleware = [thunk, enhanceReduxMiddleware];
```

And you'll get a different error message.

## Edit

```js
const middleware = enhanceReduxMiddleware([logger]);
const enhancers = [applyMiddleware(...middleware)];

// using createStore
const store = createStore(reducers, initialState, compose(...enhancers));
```

But it still outputs the error... looking into it once again

## Edit 2 :tada:

I got it to work! :tada: :tada:

It is kinda of strange, but `kepler.gl` has its own react-redux version installed, and it looks like you have this options:

- Assume the implicit `react-redux` installed by them (so you don't install it yourself, which I think is not good)
- Use something like resolutions, if you use `yarn`:
```json
  "dependencies": {
    "kepler.gl": "^2.5.5",
    "next": "12.1.5",
    "react": "17.0.2",
    "react-dom": "17.0.2",
    "react-redux": "8.0.1",
    "redux": "^4.2.0",
    "styled-components": "^5.3.5"
  },
  "resolutions": {
    "react-redux": "8.0.1"
  },
```
- `npm install react-redux@7.2.8 --save-exact` should do the trick too
  - To get the version do `npm ls react-redux` and you'll see the duplicated version

At the end of the day it is a problem with `react-redux` versions as documented here: https://github.com/reduxjs/react-redux/issues/1816

**Reply by SusantaKumar27** [reply]
Hi Joseph! You are a lifesaver. The issue was in the version. Thanks a ton, man

**Reply by icyJoseph** [reply]
Great! I think this issue is very easy to run into. 

Don't forget to mark as answered :tada:

**Reply by RedRosh** [reply]
Thanks a lot solved the issue.

**Reply by dekguh** [reply]
thanks! this answer solved my problem 🚀

**Reply by MuhammadHARIS46** [reply]
can you guys guide me the procedure if i am using redux toolkit? Next js version 13.2.3 and react version 18.2.0. I am using typescript

**Comment by dbabbs** [comment]
Has anyone had any success in integrating Kepler with Next 13 and React 18?

**Reply by MuhammadHARIS46** [reply]
no not yet

**Reply by tonalidae** [reply]
Did you find any solution?
